### ISSUE TITLE
Question about loss and training of refinement network

### ISSUE THREAD

**Issue by Enuvesta** [cvlab-stonybrook/doc3D-dataset#10]
Hi!
I have two qestions about refinement network:
1. According to the paper loss is calculated as L1 norm of difference between GT sading map and predicted shading map. Also, we can get shaded image I as hadamard product between shading free image and shading map I = A (hadamard prod) S. But there is no GT shading map in the dataset. Does it mean that you calculate ground truth S as I / A (elementwise) for training? Where I is the input image and A is corresponding albedo? 

2. Refinement network consists of two UNet-like nets. The first net takes I (input image) as an input and produces N (normals), the second one takes N and produces S (shading map). But loss contains only shading map. So, technically, this is one network and I should use one optimizer for training. Am I right?

**Comment by sagniklp** [comment]
1. Yes, that's how shading map is calculated. Make sure to scale the values of the map.
2.  The normal estimation network is pretrained with L2. While training Shading estimation we fix the weights of normal estimation.

**Comment by Enuvesta** [comment]
1. And how do you manage zero values?

**Comment by sagniklp** [comment]
Do something like this:
`shdmap=np.divide(img,alb,out=np.zeros_like(img), where=alb!=0).astype(np.float32)`

**Comment by Enuvesta** [comment]
And normalization should look like this:
`shdmap = (shdmap - shdmap.min()) / (shdmap.max() - shdmap.min())` ?

**Comment by fh2019ustc** [comment]
We print(shdmap.max(), shdmap.min()) using your code, and the results are like that:
![image](https://user-images.githubusercontent.com/50725551/106373818-5bc98d80-63b8-11eb-8134-dc01bf037753.png)
How should we scale the values? 

@Enuvesta Hi. Did you reproduce good results？

> Do something like this:
> `shdmap=np.divide(img,alb,out=np.zeros_like(img), where=alb!=0).astype(np.float32)`

**Comment by sagniklp** [comment]
We didn't scale the values. Used ReLU as the final activation for shading map regression.
However, people often use log scale.

**Comment by fh2019ustc** [comment]
@sagniklp Thanks for your reply.

For training the refinement network, the shape of the input image is 256*256, and the output of the shading map is 128*128.
For inference,  is the 128*128 shading map resized to the original shape of the distorted image, like the operation of the geometric rectification network.

Am I right? Thank you.

**Comment by sagniklp** [comment]
Yeah, correct!
